Sources/Bot/API/RepositoryAPI.swift

```diff
@@ -0,0 +1,160 @@
+    }
+
+    func pullRequest(number: UInt) -> Resource<PullRequestMetadata> {
+        return Resource(method: .GET, path: path(for: "pulls/\(number)"), decoder: decode)
+    }
+
+    func commitStatus(for pullRequest: PullRequest) -> Resource<CommitState> {
+        return Resource(
+            method: .GET,
+            path: path(for: "commits/\(pullRequest.source.sha)/status"),
+            decoder: decode
+        )
+    }
+
+    func deleteBranch(branch: PullRequest.Branch) -> Resource<Void> {
+        return Resource(
+            method: .DELETE,
+            path: path(for: "git/refs/heads/\(branch.ref)"),
+            decoder: decode
+        )
+    }
+
+    func removeLabel(label: PullRequest.Label, from pullRequest: PullRequest) -> Resource<Void> {
+        return Resource(
+            method: .DELETE,
+            path: path(for: "issues/\(pullRequest.number)/labels/\(label.name)"),
+            decoder: decode
+        )
+    }
+
+    func publish(comment: String, in pullRequest: PullRequest) -> Resource<Void> {
+        return Resource(
+            method: .POST,
+            path: path(for: "issues/\(pullRequest.number)/comments"),
+            body: encode(PostCommmentRequest(body: comment)),
+            decoder: decode
+        )
+    }
+
+    func performMerge(base: PullRequest.Branch, head: PullRequest.Branch) -> Resource<MergeResult> {
+        return Resource(
+            method: .POST,
+            path: path(for: "merges"),
+            body: encode(MergeBranchRequest(base: base, head: head)),
+            decoder: { response in
+                switch response.statusCode {
+                case 200: return .success(.success)
+                case 204: return .success(.upToDate)
+                case 409: return .success(.conflict)
+                default:
+                    return .failure(.api(response))
+                }
+            }
+        )
+    }
+
+    func merge(pullRequest: PullRequest) -> Resource<Void> {
+        return Resource(
+            method: .PUT,
+            path: "pulls/\(pullRequest.number)/merge",
+            body: encode(MergePullRequestRequest(with: pullRequest)),
+            decoder: decode
+        )
+    }
+
+    // TODO: This should be moved to be done as part of the handling of resources in `GitHubClient`
+    private func encode<T: Encodable>(_ t: T) -> Data {
+        guard let data = try? JSONEncoder().encode(t)
+            else { fatalError("Unexpected failure while encoding `\(t)`") }
+
+        return data
+    }
+}
+
+public struct RepositoryAPI: GitHubAPIProtocol {
+    private let client: GitHubClient
+    private let repository: Repository
+
+    public init(client: GitHubClient, repository: Repository) {
+        self.client = client
+        self.repository = repository
+    }
+
+    public func fetchPullRequests() -> SignalProducer<[PullRequest], AnyError> {
+        return client.request(repository.pullRequests)
+            .flatten()
+            .collect()
```

Comment:
`client.request(repository.pullRequests)` returns a `SignalProducer<[PullRequest], GitHubClient.Error>` which sends a value for each page fetched, if there's two pages we will receive two values with the contents of each page. 

As a consumer I think it's easier to receive all pull requests over multiple values representing each page, that's why we flatten everything and collect to gather everything in a single list but can be removed and then we can do it explicitly at call site.

Reply:
Ok so that's what I figured then, thanks a lot for the clarification. I'm surprised there's no single operator in RAS for that, but if not then `.flatten().collect()` makes sense 👍 
(bonus points if you add a small comment though 😛)